optimade/server/entry_collections/entry_collections.py

```diff
@@ -200,24 +203,30 @@ def _run_db_query(
         """
 
     @property
-    def all_fields(self) -> set:
+    def all_fields(self) -> Set[str]:
         """Get the set of all fields handled in this collection,
         from attribute fields in the schema, provider fields and top-level OPTIMADE fields.
 
+        The set of all fields are lazily created and then cached.
+        This means the set is created the first time the property is requested and then cached.
+
         Returns:
             All fields handled in this collection.
 
         """
-        # All OPTIMADE fields
-        fields = self.resource_mapper.TOP_LEVEL_NON_ATTRIBUTES_FIELDS.copy()
-        fields |= self.get_attribute_fields()
-        # All provider-specific fields
-        fields |= {
-            f"_{self.provider_prefix}_{field_name}"
-            for field_name in self.provider_fields
-        }
+        if not self._all_fields:
+            # All OPTIMADE fields
+            self._all_fields = (
+                self.resource_mapper.TOP_LEVEL_NON_ATTRIBUTES_FIELDS.copy()
+            )
+            self._all_fields |= self.get_attribute_fields()
+            # All provider-specific fields
+            self._all_fields |= {
+                f"_{self.provider_prefix}_{field_name}"
+                for field_name in self.provider_fields
+            }
```

Comment:
Caching looks good (we should have done this ages ago), but I still really dislike the lack of code re-use between this and the entry mapper; could you have a play with the suggestion below and see how much would need to be done to remove some of the code here? It may mean that `retrieve_queryable_properties` also needs to be updated

```suggestion
        if not self._all_fields:
            # All OPTIMADE fields
            self._all_fields = (
                self.resource_mapper.TOP_LEVEL_NON_ATTRIBUTES_FIELDS
                | set(self.resource_mapper.ENTRY_RESOURCE_ATTRIBUTES)
            )
```

Reply:
I see your point. I don't think I want to do this in this PR (although it should be fine), but a more general alignment of these classes should be made.
Also, I want to create a "handle URL parameters" class that can be sub-classed with methods matching the names of the URL query parameters :)

Reply:
> (is this fine with you or should I try to tinker?).

This is fine by me! It would also be good to fix #775...